partiql-ast/src/main/java/org/partiql/ast/literal/Literal.java

```diff
@@ -0,0 +1,125 @@
+package org.partiql.ast.literal;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.ast.AstNode;
+import org.partiql.ast.AstVisitor;
+import org.partiql.ast.DataType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Collection;
+import java.util.Collections;
+
+public abstract class Literal extends AstNode {
```

Comment:
Because of the flexibility PType provides with the fat interface, I explicitly wrote in the Javadocs that consumers of this should not author their own implementations. See https://github.com/partiql/partiql-lang-kotlin/blob/f5c6efff22543c8bb9da3fba0cb0c3cd21ea8f42/partiql-types/src/main/java/org/partiql/types/PType.java#L23-L24.

Could add that to this API as well.

Reply:
I ended up changing this to not be an abstract class and a concrete class extending `AstEnum` (like `DataType`) -- https://github.com/partiql/partiql-lang-kotlin/pull/1650#discussion_r1873778924. So adding another implementation shouldn't be an issue anymore.